README.md
# Gardian of Memory

It is said that ulimit does not work on memory limitation any more. And Linux goes crazy when memory exhausted. So made this tool.

The logic is pretty simple, scan current user's process every second. If RSS excceeds soft limit, sends libnotify notification. If RSS excceeds hard limit, sends notification and kill it. No termination yet.

# TODO

Support configurations, per command limitations.


hs/ChangeLog.md
# Changelog for GoM

## Unreleased changes


hs/app/Main.hs
{-# LANGUAGE ScopedTypeVariables #-}
module Main where

import           Control.Concurrent
import           Control.Exception
import           Control.Monad
import qualified Data.ByteString.Char8 as C8
import           Data.Char
import           GoM.Types
import           NotifySend
import           System.Directory
import           System.FilePath
import           System.Posix.Files
import           System.Posix.Signals
import           System.Posix.Types
import           System.Posix.User

main :: IO ()
main = do
  let limit0s = 393216
      limit0h = 524288
  me <- getRealUserID
  watch me limit0s limit0h

watch :: UserID -> Integer -> Integer -> IO ()
watch u limit0s limit0h = do
  listDirectory "/proc/" >>= mapM_ (\fp -> do
    isMyPid' <- maybe False id <$> wrap2Maybe (isMyPid fp u)
    when isMyPid' $ do
      mStat <- wrap2Maybe (C8.readFile ("/proc/" </> fp </> "stat")) >>=
        return . flip (>>=) readC8
      case mStat of
        Just stat -> do
          let rss = statRss stat
          if rss >= limit0s
            then if rss >= limit0h
              then kill stat
              else alert stat
            else
              return ()
        Nothing -> return ())
  threadDelay 1000000
  watch u limit0s limit0h
  where
    wrap2Maybe :: IO a -> IO (Maybe a)
    wrap2Maybe f = catch ((<$>) Just $! f) (\(_ :: IOException) -> return Nothing)
    isMyPid :: FilePath -> UserID -> IO Bool
    isMyPid fp me = do
      let areDigit = all isDigit fp
      isDir <- doesDirectoryExist $ "/proc/" </> fp
      owner <- fileOwner <$> getFileStatus ("/proc" </> fp)
      return $ areDigit && isDir && (owner == me)
    alert :: Stat -> IO ()
    alert s = do
      void $ notify "GoM" Nothing "" (show (statPid s) ++ "|" ++ take 10 (statComm s) ++ " is reaching soft limit") "" 100000
      putStrLn $ statComm s
    kill :: Stat -> IO ()
    kill s = do
      void $ notify "GoM" Nothing "" (show (statPid s) ++ "|" ++ take 10 (statComm s) ++ " is reaching hard limit") "" 100000
      putStrLn $ statComm s
      signalProcess sigKILL $ statPid s


rs/src/main.rs
extern crate notify_send;
extern crate procfs;
use handled::*;
use nix::sys::signal::{kill, SIGTERM};
use notify_send::notify;
use procfs::*;
use std::fs::{read_dir, read_to_string};
use std::os::unix::fs::MetadataExt;
use std::thread::sleep;
use std::time::Duration;
use users::{get_current_uid, uid_t};

fn main() {
    let limit0s = 393216;
    let limit0h = 524288;
    let me = get_current_uid();
    let mut handled: HandledDict = Handled::new();
    loop {
        watch(me, limit0s, limit0h, &mut handled);
        sleep(Duration::from_secs(1))
    }
}

fn watch<T: Handled>(me: uid_t, ls: i64, lh: i64, handled: &mut T) {
    for entry_ in read_dir("/proc").unwrap() {
        let _ = entry_.and_then(|entry| {
            entry.metadata().and_then(|metadata| {
                if entry.path().is_dir()
                    && entry
                        .file_name()
                        .into_string()
                        .unwrap()
                        .parse::<u64>()
                        .is_ok()
                    && metadata.uid() == me
                {
                    read_to_string(entry.path().join("stat")).and_then(|s| {
                        Ok({
                            let stat = Stat::from_str(&s)
                                .expect(&format!("Content {} cannot be parsed as stat.", &s));
                            if stat.rss >= lh && !handled.is_hard_handled(&stat) {
                                println!("{} hardlimit", stat.comm);
                                notify_(&format!(
                                    "{}|{} reached hard limit",
                                    &stat.pid, &stat.comm
                                ));
                                if let Err(e) = kill(stat.pid.pid, SIGTERM) {
                                    println!(
                                        "Killing {}|{} failed with {}",
                                        stat.pid, stat.comm, e
                                    );
                                }
                                handled.add_hard(&stat);
                            } else if stat.rss >= ls && !handled.is_soft_handled(&stat) {
                                println!("{} softlimit", stat.comm);
                                notify_(&format!("{}|{} reached soft limit", stat.pid, stat.comm));
                                handled.add_soft(&stat);
                            } else {
                                handled.del(&stat);
                            }
                        })
                    })
                } else {
                    Ok(())
                }
            })
        });
    }
}

fn notify_(msg: &str) {
    if let Err(e) = notify("GoM", None, "", msg, "", 10000) {
        println!("Sending notification faile with {}", e);
    }
}

pub mod handled {
    use procfs::*;
    pub trait Handled {
        fn new() -> Self;
        fn add_soft(&mut self, stat: &Stat) -> ();
        fn add_hard(&mut self, stat: &Stat) -> ();
        fn del(&mut self, stat: &Stat) -> ();
        fn is_soft_handled(&self, stat: &Stat) -> bool;
        fn is_hard_handled(&self, stat: &Stat) -> bool;
    }
    pub type HandledDict = (Vec<Stat>, Vec<Stat>);
    impl Handled for HandledDict {
        fn new() -> Self {
            (Vec::new(), Vec::new())
        }
        fn add_soft(&mut self, stat: &Stat) {
            self.del(stat);
            let (s, _) = self;
            if let None = s
                .iter()
                .position(|x| x.pid == stat.pid && x.comm == stat.comm)
            {
                s.push((*stat).clone());
            }
        }
        fn add_hard(&mut self, stat: &Stat) {
            self.del(stat);
            let (_, h) = self;
            if let None = h
                .iter()
                .position(|x| x.pid == stat.pid && x.comm == stat.comm)
            {
                h.push((*stat).clone());
            }
        }
        fn del(&mut self, stat: &Stat) {
            let (s, h) = self;
            if let Some(i) = h
                .iter()
                .position(|x| x.pid == stat.pid && x.comm == stat.comm)
            {
                h.remove(i);
            }
            if let Some(i) = s
                .iter()
                .position(|x| x.pid == stat.pid && x.comm == stat.comm)
            {
                s.remove(i);
            }
        }
        fn is_soft_handled(&self, stat: &Stat) -> bool {
            let (s, _) = self;
            if let Some(_) = s
                .iter()
                .position(|x| x.pid == stat.pid && x.comm == stat.comm)
            {
                true
            } else {
                false
            }
        }
        fn is_hard_handled(&self, stat: &Stat) -> bool {
            let (_, h) = self;
            if let Some(_) = h
                .iter()
                .position(|x| x.pid == stat.pid && x.comm == stat.comm)
            {
                true
            } else {
                false
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::handled::*;
    use procfs::*;
    #[test]
    fn test() {
        let mut t: HandledDict = Handled::new();
        let s = Stat::from_str("30842 (cat) R 28714 30842 28714 34816 30842 4194304 106 0 0 0 0 0 0 0 20 0 1 0 6369123 5550080 55 18446744073709551615 93870926708736 93870926733201 140735431914848 0 0 0 0 0 0 0 0 0 17 2 0 0 0 0 0 93870926748368 93870926749888 93870947766272 140735431920918 140735431920938 140735431920938 140735431925739 0").unwrap();
        t.add_soft(&s);
        assert_eq!(t.is_soft_handled(&s), true);
        assert_eq!(t.is_hard_handled(&s), false);
        t.add_hard(&s);
        assert_eq!(t.is_soft_handled(&s), false);
        assert_eq!(t.is_hard_handled(&s), true);
        t.del(&s);
        assert_eq!(t.is_soft_handled(&s), false);
        assert_eq!(t.is_hard_handled(&s), false);
    }
}


rs/src/types.rs
use nix::sys::signal::Signal;
use nix::unistd::Pid;
use num_traits::Num;
use std::fmt;
use std::num::ParseIntError;
use std::str::FromStr;

#[derive(Debug)]
pub struct StringParseError;
#[derive(Debug)]
pub struct NoneError;
impl From<ParseIntError> for StringParseError {
  fn from(_: ParseIntError) -> Self {
    StringParseError
  }
}
impl From<nix::Error> for StringParseError {
  fn from(_: nix::Error) -> Self {
    StringParseError
  }
}
impl From<NoneError> for StringParseError {
  fn from(_: NoneError) -> Self {
    StringParseError
  }
}

#[derive(Clone)]
pub struct Pid_ {
  pub pid: Pid,
}
impl FromStr for Pid_ {
  type Err = StringParseError;
  fn from_str(s: &str) -> Result<Self, Self::Err> {
    Ok(Pid_ {
      pid: Pid::from_raw(i32::from_str(s)?),
    })
  }
}
impl fmt::Display for Pid_ {
  fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
    self.pid.fmt(f)
  }
}
impl std::cmp::PartialEq for Pid_ {
  fn eq(&self, other: &Self) -> bool {
    self.pid == other.pid
  }
}
impl std::cmp::Eq for Pid_ {}

#[derive(Clone)]
pub enum ProcStat {
  PSRunning,
  PSSleeping,
  PSWaiting,
  PSZombie,
  PSStopped,
  PSTracing,
  PSPaging,
  PSDead,
  PSWakekill,
  PSWaking,
  PSParked,
}
impl FromStr for ProcStat {
  type Err = StringParseError;
  fn from_str(s: &str) -> Result<Self, Self::Err> {
    match s {
      "R" => Ok(ProcStat::PSRunning),
      "S" => Ok(ProcStat::PSSleeping),
      "D" => Ok(ProcStat::PSWaiting),
      "Z" => Ok(ProcStat::PSZombie),
      "T" => Ok(ProcStat::PSStopped),
      "t" => Ok(ProcStat::PSTracing),
      // "W" => Ok(ProcStat::PSPaging),
      "X" => Ok(ProcStat::PSDead),
      "x" => Ok(ProcStat::PSDead),
      "K" => Ok(ProcStat::PSWakekill),
      "W" => Ok(ProcStat::PSWaking),
      "P" => Ok(ProcStat::PSParked),
      _ => Err(StringParseError),
    }
  }
}

#[derive(Clone)]
pub struct DeviceNumber {
  pub major: u8,
  pub minor: u32,
}
impl FromStr for DeviceNumber {
  type Err = StringParseError;
  fn from_str(s: &str) -> Result<Self, Self::Err> {
    let s_: Vec<_> = s.splitn(2, ':').collect();
    if s_.len() == 2 {
      let major = u8::from_str(s_[0])?;
      let minor = u32::from_str(s_[1])?;
      Ok(DeviceNumber {
        major: major,
        minor: minor,
      })
    } else {
      let n = u32::from_str(s)?;
      let major = (n >> 8) & 255;
      let minor_l = n & 255;
      let minor_h = (n >> 19) & 8191;
      let minor = (minor_h << 8) | minor_l;
      Ok(DeviceNumber {
        major: major as u8,
        minor: minor,
      })
    }
  }
}

bitflags! {
  pub struct ProcessFlag: u32 {
    const PFIDLE = 0x00000002;
    const PFEXITING = 0x00000004;
    const PFEXITPIDONE = 0x00000008;
    const PFVCPU = 0x00000010;
    const PFWQWORKER = 0x00000020;
    const PFFORKNOEXEC = 0x00000040;
    const PFMCEPROCESS = 0x00000080;
    const PFSUPERPRIV = 0x00000100;
    const PFDUMPCORE = 0x00000200;
    const PFSIGNALED = 0x00000400;
    const PFMEMALLOC = 0x00000800;
    const PFNPROCEXCEEDED = 0x00001000;
    const PFUSEDMATH = 0x00002000;
    const PFUSEDASYNC = 0x00004000;
    const PFNOFREEZE = 0x00008000;
    const PFFROZEN = 0x00010000;
    const PFKSWAPD = 0x00020000;
    const PFMEMALLOCNOFS = 0x00040000;
    const PFMEMALLOCNOIO = 0x00080000;
    const PFLESSTHROTTLE = 0x00100000;
    const PFKTHREAD = 0x00200000;
    const PFRANDOMIZE = 0x00400000;
    const PFSWAPWRITE = 0x00800000;
    const PFMEMSTALL = 0x01000000;
    const PFUMH = 0x02000000;
    const PFNOSETAFFINITY = 0x04000000;
    const PFMCEEARLY = 0x08000000;
    const PFMUTEXTESTER = 0x20000000;
    const PFFREEZERSKIP = 0x40000000;
    const PFSUSPENDTASK = 0x80000000;
  }
}
impl FromStr for ProcessFlag {
  type Err = StringParseError;
  fn from_str(s: &str) -> Result<Self, Self::Err> {
    Ok(o2r(ProcessFlag::from_bits(u32::from_str(s)?))?)
  }
}

#[derive(Clone)]
pub enum SchedPolicy {
  SPNormal = 0,
  SPFifo = 1,
  SPRr = 2,
  SPBatch = 3,
  SPIso = 4,
  SPIdle = 5,
  SPDeadline = 6,
}
impl FromStr for SchedPolicy {
  type Err = StringParseError;
  fn from_str(s: &str) -> Result<Self, Self::Err> {
    match s {
      "0" => Ok(SchedPolicy::SPNormal),
      "1" => Ok(SchedPolicy::SPFifo),
      "2" => Ok(SchedPolicy::SPRr),
      "3" => Ok(SchedPolicy::SPBatch),
      "4" => Ok(SchedPolicy::SPIso),
      "5" => Ok(SchedPolicy::SPIdle),
      "6" => Ok(SchedPolicy::SPDeadline),
      _ => Err(StringParseError),
    }
  }
}

pub enum ExitStatus {
  ESExited(u8),
  ESSignaled(Signal),
  ESCoreDump,
  ESStopped(Signal),
  ESContinued,
}
impl FromStr for ExitStatus {
  type Err = StringParseError;
  fn from_str(s: &str) -> Result<Self, Self::Err> {
    let i = u64::from_str(s)?;
    let termsig = i & 0x7f;
    let exited = termsig == 0;
    let exitstatus = (i & 0xff00) >> 8;
    let signaled = (termsig + 1) >> 1 > 0;
    let coredump = i & 0x80 != 0;
    let stopped = i & 0xff == 0x7f;
    let stopsig = exitstatus;
    let continued = i == 0xffff;
    if exited {
      Ok(ExitStatus::ESExited(exitstatus as u8))
    } else if signaled {
      Ok(ExitStatus::ESSignaled(Signal::from_c_int(termsig as i32)?))
    } else if coredump {
      Ok(ExitStatus::ESCoreDump)
    } else if stopped {
      Ok(ExitStatus::ESStopped(Signal::from_c_int(stopsig as i32)?))
    } else if continued {
      Ok(ExitStatus::ESContinued)
    } else {
      Err(StringParseError)
    }
  }
}

pub type ClockTicksU = u64;
pub type ClockTicksI = i64;
pub type Address = u64;
pub type CPUId = u32;
pub type ProcessSessionId = Pid_;
pub type ProcessGroupId = Pid_;

fn from_ptrace_protected<T: Num>(s: T) -> Option<T> {
  if s.is_zero() {
    None
  } else {
    Some(s)
  }
}

fn o2r<T>(o: Option<T>) -> Result<T, NoneError> {
  if let Some(x) = o {
    Ok(x)
  } else {
    Err(NoneError)
  }
}

fn nat_or_none<U, E, F: FnOnce(i32) -> Result<U, E>>(n: i32, f: F) -> Result<Option<U>, E> {
  Ok(if n > 0 { Some(f(n)?) } else { None })
}

fn drop<Iter: Iterator>(iter: &mut Iter) -> () {
  iter.next();
}

// %d (%s) %c %d %d %d %d %d %u %lu
// %lu %lu %lu %lu %lu %ld %ld %ld %ld %ld
// %ld %llu %lu %ld %lu %lu %lu %lu %lu %lu
// %lu %lu %lu %lu %lu %lu %lu %d %d $u
// %u %llu %lu %ld %lu %lu %lu %lu %lu %lu
// %lu %d
#[derive(Clone)]
pub struct Stat {
  pub pid: Pid_,
  pub comm: String,
  pub state: ProcStat,
  pub ppid: Pid_,
  pub pgrp: ProcessGroupId,
  pub session: ProcessSessionId,
  pub tty_nr: DeviceNumber,
  pub tpgid: Option<ProcessGroupId>,
  pub flags: ProcessFlag,
  pub minflt: u64,

  pub cminflt: u64,
  pub majflt: u64,
  pub cmajflt: u64,
  pub utime: ClockTicksU,
  pub stime: ClockTicksU,
  pub cutime: ClockTicksI,
  pub cstime: ClockTicksI,
  pub priority: i64,
  pub nice: i64,
  pub num_threads: i64,

  pub itrealvalue: (),
  pub starttime: ClockTicksU,
  pub vsize: u64,
  pub rss: i64,
  pub rsslim: u64,
  pub startcode: Option<Address>,
  pub endcode: Option<Address>,
  pub startstack: Option<Address>,
  pub kstkesp: Option<Address>,
  pub kstkeip: Option<Address>,

  pub signal: (),
  pub blocked: (),
  pub sigignore: (),
  pub sigcatch: (),
  pub wchan: Option<u64>,
  pub nswap: (),
  pub cnswap: (),
  pub exit_signal: Signal,
  pub processer: CPUId,
  pub rt_priority: u32,

  pub policy: SchedPolicy,
  pub delayacct_blkio_ticks: ClockTicksU,
  pub guest_time: ClockTicksU,
  pub cguest_time: ClockTicksI,
  pub start_data: Option<Address>,
  pub end_data: Option<Address>,
  pub start_brk: Option<Address>,
  pub arg_start: Option<Address>,
  pub arg_end: Option<Address>,
  pub env_start: Option<Address>,

  pub env_end: Option<Address>,
  pub exit_code: Option<i32>,
}
impl FromStr for Stat {
  type Err = StringParseError;
  fn from_str(s: &str) -> Result<Self, Self::Err> {
    let lp = o2r(s.find('('))?;
    let rp = o2r(s.rfind(')'))?;
    let mut parts = s[(rp + 2)..].split(' ');
    Ok(Stat {
      pid: Pid_::from_str(&s[0..(lp - 1)])?,
      comm: s[(lp + 1)..rp].to_string(),
      state: ProcStat::from_str(o2r(parts.next())?)?,
      ppid: Pid_::from_str(o2r(parts.next())?)?,
      pgrp: Pid_::from_str(o2r(parts.next())?)?,
      session: Pid_::from_str(o2r(parts.next())?)?,
      tty_nr: DeviceNumber::from_str(o2r(parts.next())?)?,
      tpgid: nat_or_none::<_, StringParseError, _>(i32::from_str(o2r(parts.next())?)?, |i| {
        Ok(Pid_ {
          pid: Pid::from_raw(i),
        })
      })?,
      flags: ProcessFlag::from_str(o2r(parts.next())?)?,
      minflt: u64::from_str(o2r(parts.next())?)?,

      cminflt: u64::from_str(o2r(parts.next())?)?,
      majflt: u64::from_str(o2r(parts.next())?)?,
      cmajflt: u64::from_str(o2r(parts.next())?)?,
      utime: u64::from_str(o2r(parts.next())?)?,
      stime: u64::from_str(o2r(parts.next())?)?,
      cutime: i64::from_str(o2r(parts.next())?)?,
      cstime: i64::from_str(o2r(parts.next())?)?,
      priority: i64::from_str(o2r(parts.next())?)?,
      nice: i64::from_str(o2r(parts.next())?)?,
      num_threads: i64::from_str(o2r(parts.next())?)?,

      itrealvalue: drop(&mut parts),
      starttime: u64::from_str(o2r(parts.next())?)?,
      vsize: u64::from_str(o2r(parts.next())?)?,
      rss: i64::from_str(o2r(parts.next())?)?,
      rsslim: u64::from_str(o2r(parts.next())?)?,
      startcode: from_ptrace_protected(u64::from_str(o2r(parts.next())?)?),
      endcode: from_ptrace_protected(u64::from_str(o2r(parts.next())?)?),
      startstack: from_ptrace_protected(u64::from_str(o2r(parts.next())?)?),
      kstkesp: from_ptrace_protected(u64::from_str(o2r(parts.next())?)?),
      kstkeip: from_ptrace_protected(u64::from_str(o2r(parts.next())?)?),

      signal: drop(&mut parts),
      blocked: drop(&mut parts),
      sigignore: drop(&mut parts),
      sigcatch: drop(&mut parts),
      wchan: from_ptrace_protected(u64::from_str(o2r(parts.next())?)?),
      nswap: drop(&mut parts),
      cnswap: drop(&mut parts),
      exit_signal: Signal::from_c_int(i32::from_str(o2r(parts.next())?)?)?,
      processer: u32::from_str(o2r(parts.next())?)?,
      rt_priority: u32::from_str(o2r(parts.next())?)?,

      policy: SchedPolicy::from_str(o2r(parts.next())?)?,
      delayacct_blkio_ticks: u64::from_str(o2r(parts.next())?)?,
      guest_time: u64::from_str(o2r(parts.next())?)?,
      cguest_time: i64::from_str(o2r(parts.next())?)?,
      start_data: from_ptrace_protected(u64::from_str(o2r(parts.next())?)?),
      end_data: from_ptrace_protected(u64::from_str(o2r(parts.next())?)?),
      start_brk: from_ptrace_protected(u64::from_str(o2r(parts.next())?)?),
      arg_start: from_ptrace_protected(u64::from_str(o2r(parts.next())?)?),
      arg_end: from_ptrace_protected(u64::from_str(o2r(parts.next())?)?),
      env_start: from_ptrace_protected(u64::from_str(o2r(parts.next())?)?),

      env_end: from_ptrace_protected(u64::from_str(o2r(parts.next())?)?),
      exit_code: from_ptrace_protected(i32::from_str(o2r(parts.next())?.trim_end())?),
    })
  }
}


hs/src/GoM/Types.hs
{-# LANGUAGE TypeSynonymInstances #-}
module GoM.Types where

import           Data.Bits
import qualified Data.ByteString.Char8 as C8
import           Data.Word
import           GoM.Helper
import           System.Exit
import           System.Posix.Signals
import           System.Posix.Types

class ReadC8 a where
  readC8 :: C8.ByteString -> Maybe a

data ProcState = PSRunning
               | PSSleeping
               | PSWaiting
               | PSZombie
               | PSStopped
               | PSTracing
               | PSPaging
               | PSDead
               | PSWakekill
               | PSWaking
               | PSParked
               deriving (Show)
instance ReadC8 ProcState where
  readC8 s = lookup s mapping
    where
      mapping :: [(C8.ByteString, ProcState)]
      mapping = [ (C8.singleton 'R', PSRunning)
                , (C8.singleton 'S', PSSleeping)
                , (C8.singleton 'D', PSWaiting)
                , (C8.singleton 'Z', PSZombie)
                , (C8.singleton 'T', PSStopped)
                , (C8.singleton 't', PSTracing)
--                , (C8.singleton 'W', PSPaging)
                , (C8.singleton 'X', PSDead)
                , (C8.singleton 'x', PSDead)
                , (C8.singleton 'K', PSWakekill)
                , (C8.singleton 'W', PSWaking)
                , (C8.singleton 'P', PSParked)]

data DeviceNumber = DN { dnMajor :: Word8
                       , dnMinor :: Word32 }
                  deriving (Show)
instance ReadC8 DeviceNumber where
  readC8 s = do
    (n, r) <- C8.readInteger s
    if C8.empty == r
      then do
        let major = shiftR n 8 .&. 255
            minorL = n .&. 255
            minorH = shiftR n 19 .&. 8191
            minor = shiftL minorH 8 .|. minorL
        return $ DN (fromIntegral major) (fromIntegral minor)
      else if C8.length r > 1 && C8.head r == ':'
      then do
        m <- readIntegerAll $ C8.tail r
        return $ DN (fromIntegral n) (fromIntegral m)
      else fail ""

data ProcessFlag = PFIdle
                 | PFExiting
                 | PFExitpidone
                 | PFVcpu
                 | PFWqWorker
                 | PFForknoexec
                 | PFMceProcess
                 | PFSuperpriv
                 | PFDumpcore
                 | PFSignaled
                 | PFMemalloc
                 | PFNprocExceeded
                 | PFUsedMath
                 | PFUsedAsync
                 | PFNofreeze
                 | PFFrozen
                 | PFKswapd
                 | PFMemallocNofs
                 | PFMemallocNoio
                 | PFLessThrottle
                 | PFKthread
                 | PFRandomize
                 | PFSwapwrite
                 | PFMemstall
                 | PFUmh
                 | PFNoSetaffinity
                 | PFMceEarly
                 | PFMutexTester
                 | PFFreezerSkip
                 | PFSuspendTask
                 deriving (Show)
instance Enum ProcessFlag where
  fromEnum PFIdle = 0x00000002
  fromEnum PFExiting = 0x00000004
  fromEnum PFExitpidone = 0x00000008
  fromEnum PFVcpu = 0x00000010
  fromEnum PFWqWorker = 0x00000020
  fromEnum PFForknoexec = 0x00000040
  fromEnum PFMceProcess = 0x00000080
  fromEnum PFSuperpriv = 0x00000100
  fromEnum PFDumpcore = 0x00000200
  fromEnum PFSignaled = 0x00000400
  fromEnum PFMemalloc = 0x00000800
  fromEnum PFNprocExceeded = 0x00001000
  fromEnum PFUsedMath = 0x00002000
  fromEnum PFUsedAsync = 0x00004000
  fromEnum PFNofreeze = 0x00008000
  fromEnum PFFrozen = 0x00010000
  fromEnum PFKswapd = 0x00020000
  fromEnum PFMemallocNofs = 0x00040000
  fromEnum PFMemallocNoio = 0x00080000
  fromEnum PFLessThrottle = 0x00100000
  fromEnum PFKthread = 0x00200000
  fromEnum PFRandomize = 0x00400000
  fromEnum PFSwapwrite = 0x00800000
  fromEnum PFMemstall = 0x01000000
  fromEnum PFUmh = 0x02000000
  fromEnum PFNoSetaffinity = 0x04000000
  fromEnum PFMceEarly = 0x08000000
  fromEnum PFMutexTester = 0x20000000
  fromEnum PFFreezerSkip = 0x40000000
  fromEnum PFSuspendTask = 0x80000000
  toEnum 0x00000002 = PFIdle
  toEnum 0x00000004 = PFExiting
  toEnum 0x00000008 = PFExitpidone
  toEnum 0x00000010 = PFVcpu
  toEnum 0x00000020 = PFWqWorker
  toEnum 0x00000040 = PFForknoexec
  toEnum 0x00000080 = PFMceProcess
  toEnum 0x00000100 = PFSuperpriv
  toEnum 0x00000200 = PFDumpcore
  toEnum 0x00000400 = PFSignaled
  toEnum 0x00000800 = PFMemalloc
  toEnum 0x00001000 = PFNprocExceeded
  toEnum 0x00002000 = PFUsedMath
  toEnum 0x00004000 = PFUsedAsync
  toEnum 0x00008000 = PFNofreeze
  toEnum 0x00010000 = PFFrozen
  toEnum 0x00020000 = PFKswapd
  toEnum 0x00040000 = PFMemallocNofs
  toEnum 0x00080000 = PFMemallocNoio
  toEnum 0x00100000 = PFLessThrottle
  toEnum 0x00200000 = PFKthread
  toEnum 0x00400000 = PFRandomize
  toEnum 0x00800000 = PFSwapwrite
  toEnum 0x01000000 = PFMemstall
  toEnum 0x02000000 = PFUmh
  toEnum 0x04000000 = PFNoSetaffinity
  toEnum 0x08000000 = PFMceEarly
  toEnum 0x20000000 = PFMutexTester
  toEnum 0x40000000 = PFFreezerSkip
  toEnum 0x80000000 = PFSuspendTask
  toEnum _ = undefined
readProcessFlags :: C8.ByteString -> Maybe [ProcessFlag]
readProcessFlags flags = do
  flags' <- readIntegerAll flags
  return $ filter (\enum -> fromEnum enum .&. fromIntegral flags' == fromEnum enum) $
    map (toEnum . shiftL 1) $ [1..27] ++ [29, 30, 31]

data SchedPolicy = SPNormal
                 | SPFifo
                 | SPRr
                 | SPBatch
                 | SPIso
                 | SPIdle
                 | SPDeadline
                 deriving (Show)
instance Enum SchedPolicy where
  fromEnum SPNormal = 0
  fromEnum SPFifo = 1
  fromEnum SPRr = 2
  fromEnum SPBatch = 3
  fromEnum SPIso = 4
  fromEnum SPIdle = 5
  fromEnum SPDeadline = 6
  toEnum 0 = SPNormal
  toEnum 1 = SPFifo
  toEnum 2 = SPRr
  toEnum 3 = SPBatch
  toEnum 4 = SPIso
  toEnum 5 = SPIdle
  toEnum 6 = SPDeadline
  toEnum _ = undefined

data ExitStatus = ESExited ExitCode
                | ESSignaled Signal
                | ESCoreDump
                | ESStoppped Signal
                | ESContinued
                deriving (Eq, Show)
instance ReadC8 ExitStatus where
  readC8 s = do
    i <- readIntegerAll s
    let termsig = i .&. 0x7F
        exited = termsig == 0
        exitstatus = shiftR (i .&. 0xFF00) 8
        signaled = shiftR (termsig + 1) 1 > 0
        coredump = i .&. coreflag /= 0
        stopped = i .&. 0xFF == 0x7F
        stopsig = exitstatus
        continued = i == 0xFFFF
    if exited
      then return $ ESExited $ case exitstatus of
        0 -> ExitSuccess
        j -> ExitFailure $ fromIntegral j
      else if signaled
      then return $ ESSignaled $ fromIntegral termsig
      else if coredump
      then return ESCoreDump
      else if stopped
      then return $ ESStoppped $ fromIntegral stopsig
      else if continued
      then return ESContinued
      else fail ""
    where
      coreflag = 0x80

instance ReadC8 Int where
  readC8 = fmap fromIntegral . readIntegerAll

instance ReadC8 Signal where
  readC8 = fmap fromIntegral . readIntegerAll

instance ReadC8 Integer where
  readC8 = readIntegerAll

instance ReadC8 ProcessID where
  readC8 = fmap fromIntegral . readIntegerAll

type ProcessSessionID = CPid
type ClockTicks = Integer
type Address = Integer
type CPUId = Integer

data Stat = Stat { statPid :: ProcessID
                 , statComm :: String
                 , statState :: ProcState
                 , statPpid :: ProcessID
                 , statPgrp :: ProcessGroupID
                 , statSession :: ProcessSessionID
                 , statTtyNr :: DeviceNumber
                 , statTpgid :: Maybe ProcessGroupID
                 , statFlags :: [ProcessFlag]
                 , statMinflt :: Integer
                 , statCminflt :: Integer
                 , statMajflt :: Integer
                 , statCmajflt :: Integer
                 , statUtime :: ClockTicks
                 , statStime :: ClockTicks
                 , statCutime :: ClockTicks
                 , statCstime :: ClockTicks
                 , statPriority :: Integer
                 , statNice :: Integer
                 , statNumThreads :: Integer
                 , statItrealvalue :: ()
                 , statStarttime :: ClockTicks
                 , statVsize :: Integer
                 , statRss :: Integer
                 , statRsslim :: Integer
                 , statStartcode :: Maybe Address
                 , statEndcode :: Maybe Address
                 , statStartstack :: Maybe Address
                 , statKstkesp :: Maybe Address
                 , statKstkeip :: Maybe Address
                 , statSignal :: ()
                 , statBlocked :: ()
                 , statSigignore :: ()
                 , statSigcatch :: ()
                 , statWchan :: Maybe Integer
                 , statNswap :: ()
                 , statCnswap :: ()
                 , statExitSignal :: Signal
                 , statProcesser :: CPUId
                 , statRtPriority :: Integer
                 , statPolicy :: SchedPolicy
                 , statDelayacctBlkioTicks :: ClockTicks
                 , statGuestTime :: ClockTicks
                 , statCguestTime :: ClockTicks
                 , statStartData :: Maybe Address
                 , statEndData :: Maybe Address
                 , statStartBrk :: Maybe Address
                 , statArgStart :: Maybe Address
                 , statArgEnd :: Maybe Address
                 , statEnvStart :: Maybe Address
                 , statEnvEnd :: Maybe Address
                 , statExitCode :: Maybe ExitStatus }
          deriving (Show)
instance ReadC8 Stat where
  readC8 s = do
    lp <- C8.elemIndex '(' s
    rp <- C8.elemIndexEnd ')' s
    let pid' = C8.take (lp - 1) s
        comm' = C8.unpack $ C8.drop (lp + 1) $ C8.take rp s
        fields = [C8.empty, C8.empty] ++ C8.split ' ' (C8.drop (rp + 2) $ C8.init s)
    pure Stat
      <*> readC8 pid'
      <*> pure comm'
      <*> readC8 (fields !! 2)
      <*> readC8 (fields !! 3)
      <*> readC8 (fields !! 4)
      <*> readC8 (fields !! 5)
      <*> readC8 (fields !! 6)
      <*> fmap (nothingOnValue (-1)) (readC8 (fields !! 7))
      <*> readProcessFlags (fields !! 8)
      <*> readC8 (fields !! 9)
      <*> readC8 (fields !! 10)
      <*> readC8 (fields !! 11)
      <*> readC8 (fields !! 12)
      <*> readC8 (fields !! 13)
      <*> readC8 (fields !! 14)
      <*> readC8 (fields !! 15)
      <*> readC8 (fields !! 16)
      <*> readC8 (fields !! 17)
      <*> readC8 (fields !! 18)
      <*> readC8 (fields !! 19)
      <*> pure ()
      <*> readC8 (fields !! 21)
      <*> readC8 (fields !! 22)
      <*> readC8 (fields !! 23)
      <*> readC8 (fields !! 24)
      <*> fmap (nothingOnValue 0) (readC8 (fields !! 25))
      <*> fmap (nothingOnValue 0) (readC8 (fields !! 26))
      <*> fmap (nothingOnValue 0) (readC8 (fields !! 27))
      <*> fmap (nothingOnValue 0) (readC8 (fields !! 28))
      <*> fmap (nothingOnValue 0) (readC8 (fields !! 29))
      <*> pure ()
      <*> pure ()
      <*> pure ()
      <*> pure ()
      <*> fmap (nothingOnValue 0) (readC8 (fields !! 34))
      <*> pure ()
      <*> pure ()
      <*> readC8 (fields !! 37)
      <*> readC8 (fields !! 38)
      <*> readC8 (fields !! 39)
      <*> (toEnum <$> readC8 (fields !! 40))
      <*> readC8 (fields !! 41)
      <*> readC8 (fields !! 42)
      <*> readC8 (fields !! 43)
      <*> fmap (nothingOnValue 0) (readC8 (fields !! 44))
      <*> fmap (nothingOnValue 0) (readC8 (fields !! 45))
      <*> fmap (nothingOnValue 0) (readC8 (fields !! 46))
      <*> fmap (nothingOnValue 0) (readC8 (fields !! 47))
      <*> fmap (nothingOnValue 0) (readC8 (fields !! 48))
      <*> fmap (nothingOnValue 0) (readC8 (fields !! 49))
      <*> fmap (nothingOnValue 0) (readC8 (fields !! 50))
      <*> fmap (nothingOnValue (ESExited ExitSuccess)) (readC8 (fields !! 51))
      where
        nothingOnValue :: (Eq a) => a -> a -> Maybe a
        nothingOnValue x y = if x == y
          then Nothing
          else Just y
